hw002_MapsGenerics/src/task7/main/main.go

```diff
@@ -0,0 +1,134 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+)
+
+type Exam struct {
+	Students []Student `json:"students"`
+	Objects  []Object  `json:"objects"`
+	Results  []Result  `json:"results"`
+}
+type Student struct {
+	Id    int    `json:"id"`
+	Name  string `json:"name"`
+	Grade int    `json:"grade"`
+}
+type Object struct {
+	Id   int    `json:"id"`
+	Name string `json:"name"`
+}
+type Result struct {
+	ObjectId  int `json:"object_id"`
+	StudentId int `json:"student_id"`
+	Res       int `json:"result"`
+}
+type ObjectStatistic struct {
+	AllResult int
+	Objects   []int
+}
+
+func main() {
+	exam, err := parseJson("src/dz3.json")
+	if err != nil {
+		return
+	}
+	if len(exam.Students) == 0 {
+		fmt.Println("Файл не содержит данных")
+		return
+	}
+
+	mapStudents := getMapStudents(exam.Students)
+	mapObjects := getMapObjects(exam.Objects)
+
+	fmt.Printf("%-44s\n", strings.Repeat("_", 44))
+	fmt.Printf("%-14s|%-7s|%-11s|%-9s\n", " Student name", " Grade", " Object", " Result")
+	fmt.Printf("%-44s\n", strings.Repeat("_", 44))
+
+	mapBalls := make(map[int]ObjectStatistic)
+
+	for _, r := range exam.Results {
+		statistic := mapBalls[r.StudentId]
+		statistic.AllResult += r.Res
+		statistic.Objects = append(statistic.Objects, r.ObjectId)
+		mapBalls[r.StudentId] = statistic
+	}
+
+	excellentStudents := Filter(mapBalls, func(i int) bool {
+		if mapBalls[i].AllResult/len(mapBalls[i].Objects) == 5 {
+			return true
+		}
+		return false
+	})
+	for studentId, _ := range excellentStudents {
+		student := findStudentById(mapStudents, studentId)
+		studentName := student.Name
+		studentGrade := student.Grade
+		objectStatistic := excellentStudents[studentId]
+		for _, idObject := range objectStatistic.Objects {
+			object := findObjectById(mapObjects, idObject)
+			objectName := object.Name
+			fmt.Printf(" %-13s|%5d  | %-10s|  %-7d\n", studentName, studentGrade, objectName, 5)
+		}
+	}
+}
+
+func Filter[T any](s map[int]T, f func(int) bool) map[int]T {
+	m := make(map[int]T)
+	for i, v := range s {
+		if f(i) {
+			m[i] = v
+		}
+	}
+	return m
+}
```

Comment:
здорово, что переписала функцию под мапу
но для полноценного дженерика, я думаю, лучше было бы передавать не int как ключ мапы, а еще один параметр типа, тогда была бы вообще красота)
```suggestion
func Filter[K comparable, V any](s map[K]V, f func(K) bool) map[K]V {
	m := make(map[K]V)
	for k, v := range s {
		if f(k) {
			m[k] = v
		}
	}
	return m
}
```

Reply:
и лучше не использовать переменную под названием i как ключ мапы, у нее есть общепринятые k, v.
А i - обычно все-таки обозначает индекс